### PR TITLE
Extend catchpoint file download time

### PR DESCRIPTION
**Pull request by tsachiherman**
## Summary

This PR adjusts the time limitation of the catchpoint file download to be 2 hours, and make it a configurable via the config.json file. It also makes the minimal catchpoint file download speed to be configurable via the config.json.

## Test Plan

Unit tests were updated

### CODE REVIEW

File: config/config.go [hunk 1]
```diff
@@ -349,6 +349,16 @@ type Local struct {
 	// and is expected to happen only on either startup ( after enabling the catchpoint interval, or on certain database upgrades ) or during fast catchup. The values specified here
 	// and their meanings are identical to the ones in LedgerSynchronousMode.
 	AccountsRebuildSynchronousMode int `version[12]:"1"`
+
+	// MaxCatchpointDownloadTimeSec defines the maximum amount of time a client will be keeping the outgoing connection of a catchpoint download request open for processing before
+	// shutting it down. Networks that has large catchpoint files, slow connection or slow storage could be a good reason to increase this value. Note that this is a client-side only
```
Comment (algonautshant) [inline config/config.go:354]: typo: Networks that **have**

File: config/config.go [hunk 2]
```diff
@@ -349,6 +349,16 @@ type Local struct {
...
+	// configuration value, and it's independent of the actual catchpoint file size.
+	MaxCatchpointDownloadTimeSec uint64 `version[13]:"7200"`
+
+	// MinCatchpointFileDownloadRate defines the minimal download speed that would considered to be "acceptable" by the catchpoint file fetcher, measured in bytes per seconds. If the
```
Comment (algonautshant) [inline config/config.go:358]: type: that would **be** considered

File: config/config.go [hunk 3]
```diff
@@ -349,6 +349,16 @@ type Local struct {
...
+	// provided stream speed drops below this threshold, the connection would be recycled. Note that this field is evaluated per catchpoint "chunk" and not on it's own. This field should
```
Comment (algonautshant) [inline config/config.go:359]: Is this statement:
`This field should always be non-zero, or the default of 20480 would be used.`
the same as:
`If this field is zero, the default of 20480 would be used.`

File: config/config.go [hunk 4]
```diff
@@ -349,6 +349,16 @@ type Local struct {
...
+	// always be non-zero, or the default of 20480 would be used.
+	MinCatchpointFileDownloadRate uint64 `version[13]:"20480"`
```
Comment (algonautshant) [inline config/config.go:361]: To be consistent with the server, and eliminate the ambiguity of the units, I prefer to call this MinCatchpointFileDownloadBytesPerSecond.

File: catchup/ledgerFetcher.go [hunk 5]
```diff
@@ -39,12 +40,8 @@ var errNoLedgerForRound = errors.New("No ledger available for given round")
...
 	// maxCatchpointFileChunkSize is a rough estimate for the worst-case scenario we're going to have of all the accounts data per a single catchpoint file chunk.
 	maxCatchpointFileChunkSize = ledger.BalancesPerCatchpointFileChunk * basics.MaxEncodedAccountDataSize
-	// maxCatchpointFileDownloadDuration is the maximum time we would wait for download an entire catchpoint file
-	maxCatchpointFileDownloadDuration = 45 * time.Minute
-	// expectedWorstDownloadSpeedBytesPerSecond defines the worst case-scenario download speed we expect to get while downloading a catchpoint file
-	expectedWorstDownloadSpeedBytesPerSecond = 20 * 1024
-	// maxCatchpointFileChunkDownloadDuration is the maximum amount of time we would wait to download a single chunk off a catchpoint file
-	maxCatchpointFileChunkDownloadDuration = 2*time.Minute + maxCatchpointFileChunkSize*time.Second/expectedWorstDownloadSpeedBytesPerSecond
+	// defaultMinCatchpointFileDownloadRate defines the worst case-scenario download speed we expect to get while downloading a catchpoint file
+	defaultMinCatchpointFileDownloadRate = 20 * 1024
```
Comment (algonautshant) [inline catchup/ledgerFetcher.go:44]: To be consistent with the server, and eliminate the ambiguity of the units, I prefer to call this defaultMinCatchpointFileDownloadBytesPerSecond.

**Review by algonautshant — APPROVED** [review]
Looks good. 
A few typos and a name suggestion.

**Review by algonautshant — APPROVED** [review]
Thanks for the updates.

File: catchup/ledgerFetcher.go [hunk 6]
```diff
@@ -39,12 +40,8 @@ var errNoLedgerForRound = errors.New("No ledger available for given round")
 const (
 	// maxCatchpointFileChunkSize is a rough estimate for the worst-case scenario we're going to have of all the accounts data per a single catchpoint file chunk.
 	maxCatchpointFileChunkSize = ledger.BalancesPerCatchpointFileChunk * basics.MaxEncodedAccountDataSize
-	// maxCatchpointFileDownloadDuration is the maximum time we would wait for download an entire catchpoint file
-	maxCatchpointFileDownloadDuration = 45 * time.Minute
-	// expectedWorstDownloadSpeedBytesPerSecond defines the worst case-scenario download speed we expect to get while downloading a catchpoint file
-	expectedWorstDownloadSpeedBytesPerSecond = 20 * 1024
-	// maxCatchpointFileChunkDownloadDuration is the maximum amount of time we would wait to download a single chunk off a catchpoint file
-	maxCatchpointFileChunkDownloadDuration = 2*time.Minute + maxCatchpointFileChunkSize*time.Second/expectedWorstDownloadSpeedBytesPerSecond
+	// defaultMinCatchpointFileDownloadBytesPerSecond defines the worst case-scenario download speed we expect to get while downloading a catchpoint file
```
Comment (nicholasguoalgorand) [inline catchup/ledgerFetcher.go:43]: nit: should be "worst-case scenario"

File: config/config.go [hunk 7]
```diff
@@ -349,6 +349,16 @@ type Local struct {
 	// and is expected to happen only on either startup ( after enabling the catchpoint interval, or on certain database upgrades ) or during fast catchup. The values specified here
 	// and their meanings are identical to the ones in LedgerSynchronousMode.
 	AccountsRebuildSynchronousMode int `version[12]:"1"`
+
+	// MaxCatchpointDownloadTimeSec defines the maximum amount of time a client will be keeping the outgoing connection of a catchpoint download request open for processing before
+	// shutting it down. Networks that have large catchpoint files, slow connection or slow storage could be a good reason to increase this value. Note that this is a client-side only
+	// configuration value, and it's independent of the actual catchpoint file size.
+	MaxCatchpointDownloadTimeSec uint64 `version[13]:"7200"`
```
Comment (nicholasguoalgorand) [inline config/config.go:356]: why not have this be a time.Duration?

**Review by nicholasguoalgorand — APPROVED** [review]
Had a couple nits, otherwise LGTM

File: config/config.go (hunk 7 above)
Comment (tsachiherman) [reply config/config.go]: sounds like a good idea.

**Review by nicholasguoalgorand — APPROVED** [review]

**Pull request merged**